runtime/memprof.c

```diff
@@ -289,44 +285,38 @@ static void purge_postponed_queue(void)
    block is allocated, but not yet initialized, so that the heap
    invariants are broken. */
 static void register_postponed_callback(value block, uintnat occurrences,
-                                        enum ml_alloc_kind kind)
+                                        enum ml_alloc_kind kind,
+                                        value* callstack)
 {
-  value callstack;
   struct postponed_block* new_hd;
   if (occurrences == 0) return;
-  callstack = capture_callstack_postponed();
-  if (callstack == 0) return;    /* OOM */
+  if (*callstack == 0) *callstack = capture_callstack_postponed();
+  if (*callstack == 0) return;    /* OOM */
 
-  new_hd = postponed_hd + 1;
-  if (new_hd == postponed_queue_end) new_hd = postponed_queue;
+  new_hd = postponed_next(postponed_hd);
   if (new_hd == postponed_tl) {
     /* Queue is full, reallocate it. (We always leave one free slot in
        order to be able to distinguish the 100% full and the empty
        states). */
-    uintnat sz = 4 * (postponed_queue_end - postponed_queue);
+    uintnat
+      sz1 = postponed_queue_end - new_hd,
```

Comment:
Is this right? It seems to include the uninitialised empty slot at `postponed_hd` in `sz1`, which gets copied below.